controllers/pipelinegates/pipelinegate_controller.go

```diff
@@ -0,0 +1,376 @@
+		}
+	}
+
+	// PUSH SYNC
+	if gate.Status.State != nil &&
+		(gate.Status.LastReported == nil || ( // never reportedd
+		// the last sync from console happened AFTER we last reported the state, meaning we have not already reported the new state to the console, we do this check to avoid spamming the console with updates
+		gate.Status.LastReportedAt != nil &&
+			gate.Spec.LastSyncedAt != nil &&
+			(*gate.Status.LastReportedAt).Before(gate.Spec.LastSyncedAt))) &&
+		(*gate.Status.State == pipelinesv1alpha1.GateState(console.GateStateOpen) || *gate.Status.State == pipelinesv1alpha1.GateState(console.GateStateClosed)) {
+
+		log.V(1).Info(fmt.Sprintf("Reconciling %s gate.", string(*gate.Status.State)), "Name", gate.Name, "ID", gate.Spec.ID, "State", *gate.Status.State, "LastSyncedAt", gate.Spec.LastSyncedAt, "jobRef", *gate.Status.JobRef)
+
+		// def state var for update
+		var gateState console.GateState
+		if *gate.Status.State == pipelinesv1alpha1.GateState(console.GateStateOpen) {
+			gateState = console.GateStateOpen
+		} else {
+			gateState = console.GateStateClosed
+		}
+
+		updateAttributes := console.GateUpdateAttributes{State: &gateState, Status: &console.GateStatusAttributes{JobRef: gate.Status.JobRef}}
+
+		// try to update gate at console
+		if err := r.ConsoleClient.UpdateGate(gate.Spec.ID, updateAttributes); err != nil {
+			log.Error(err, "Failed to update PipelineGate status to console")
+			return ctrl.Result{}, err
+		}
+
+		lastReportedAt := metav1.Now()
+		lastReported := pipelinesv1alpha1.GateState(gateState)
+		log.V(2).Info(fmt.Sprintf("Updated gate state to %s console.", string(*gate.Status.State)), "Name", gate.Name, "ID", gate.Spec.ID, "State", *gate.Status.State, "lastReported", lastReported, "LastReportedAt", lastReportedAt, "LastSyncedAt", gate.Spec.LastSyncedAt, "jobRef", *gate.Status.JobRef)
+
+		// update CR state
+		gate.Status.LastReportedAt = &lastReportedAt
+		gate.Status.LastReported = &lastReported
+		if err := r.Status().Update(ctx, gate); err != nil {
+			log.Error(err, "Failed to update PipelineGate status at CR")
+			return ctrl.Result{}, err
+		}
+
+		log.V(2).Info(fmt.Sprintf("Updated gate state lastReported to %s.", string(lastReported)), "Name", gate.Name, "ID", gate.Spec.ID, "State", *gate.Status.State, "LastReportedAt", *gate.Status.LastReportedAt, "LastSyncedAt", gate.Spec.LastSyncedAt, "jobRef", *gate.Status.JobRef)
+
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func hasFailed(job *batchv1.Job) (bool, *batchv1.JobCondition) {
+	conditions := job.Status.Conditions
+	// check if the conditions slice contains a failed condition
+	// failed means the backoff limit was reached and it is not being retried anymore!
+	for _, condition := range conditions {
+		if condition.Type == batchv1.JobFailed {
+			return true, &condition
+		}
+	}
+	return false, nil
+}
+
+func hasSucceeded(job *batchv1.Job) (bool, *batchv1.JobCondition) {
+	conditions := job.Status.Conditions
```

Comment:
You can use `k8s.io/apimachinery/pkg/api/meta.IsStatusConditionPresentAndEqual` or other methods from this package. It can replace `hasFailed` and `hasSucceeded` functions. At the moment you only check if there is a condition on a job, but do not check its status (`true` or `false`).

Reply:
I coulnd't find a method in that package or elsewhere that works with JobConditions, they seem to have their own struct types (for whatever reason), I will amend my methods though, to account for the `true` `false` check (it SHOULD always be `true` from what I can tell by looking at the job API but you never know